app/code/Magento/Swatches/view/frontend/web/js/swatch-renderer.js

```diff
@@ -329,6 +329,28 @@ define([
         },
 
         /**
+         * Determine product id and related data
+         *
+         * @returns {{productId: *, isInProductView: boolean}}
+         * @private
+         */
+        _determineProductData: function() {
+            // Check if product is in a list of products.
+            var productId = this.element.parents('.product-item-details')
+                    .find('.price-box.price-final_price').attr('data-product-id'),
+                isInProductView = false;
+
+            if (!productId) {
+                // Check individual product.
+                var product = document.getElementsByName('product')[0];
```

Comment:
Hello @zetlen ,

Thank you for looking into this PR with so much care.

- I did refactor the code to use jQuery instead of plain Javascript for an easier read: e9c42d214f252ac7deaa3c26be118541366c137f.
- However, it's worth mentioning that `document.getElementsByName('product')[0]` doesn't return an exception because `document.getElementsByName('product')` returns empty array in case no elements are found and `document.getElementsByName('product')[0]` will return undefined.
- The previous code `document.getElementsByName('product')[0].value` was indeed suspicious to throw an exception - issue which this PR fixed.